macro/agent_population.py

```diff
@@ -0,0 +1,210 @@
+    combine_indep_dstns,
+)
+from xarray import DataArray
+
+ParameterDict = NewType("ParameterDict", dict)
+
+
+@dataclass
+class AgentPopulation:
+    agent_class: AgentType
+    parameter_dict: ParameterDict
+    t_age: int = None
+    agent_class_count: int = None
+
+    def __post_init__(self):
+
+        self.time_var = self.agent_class.time_vary
+        self.time_inv = self.agent_class.time_inv
+
+        self.infer_counts()
+
+    def infer_counts(self):
+
+        param_dict = self.parameter_dict
+
+        # if agent_clas_count is not specified, infer from parameters
+        if self.agent_class_count is None:
+
+            agent_class_count = 1
+            for key_param in param_dict:
+                parameter = param_dict[key_param]
+                if isinstance(parameter, DataArray) and parameter.dims[0] == "agent":
+                    agent_class_count = max(agent_class_count, parameter.shape[0])
+                elif isinstance(parameter, (Distribution, IndexDistribution)):
+                    agent_class_count = None
+                    break
+
+            self.agent_class_count = agent_class_count
+
+        if self.t_age is None:
+
+            t_age = 1
+            for key_param in param_dict:
+                parameter = param_dict[key_param]
+                if isinstance(parameter, DataArray) and parameter.dims[-1] == "age":
+                    t_age = max(t_age, parameter.shape[-1])
+                    # there may not be a good use for this feature yet as time varying distributions
+                    # are entered as list of moments (Avg, Std, Count, etc)
+                elif isinstance(parameter, (Distribution, IndexDistribution)):
+                    t_age = None
+                    break
+            self.t_age = t_age
+
+    def approx_distributions(self, approx_params: dict):
+
+        param_dict = self.parameter_dict
+
+        self.continuous_distributions = {}
+
+        self.discrete_distributions = {}
+
+        for key in approx_params:
+            if key in param_dict and isinstance(param_dict[key], Distribution):
+                discrete_points = approx_params[key]
+                discrete_distribution = param_dict[key].approx(discrete_points)
+                self.continuous_distributions[key] = param_dict[key]
+                self.discrete_distributions[key] = discrete_distribution
+            else:
+                print(
+                    "Warning: parameter {} is not a Distribution found in agent class {}".format(
+                        key, self.agent_class
+                    )
+                )
+
+        if len(self.discrete_distributions) > 1:
+            joint_dist = combine_indep_dstns(
+                *list(self.discrete_distributions.values())
+            )
+
+        keys = list(self.discrete_distributions.keys())
+        for i in range(len(self.discrete_distributions)):
+            param_dict[keys[i]] = DataArray(joint_dist.X[i], dims=("agent"))
```

Comment:
This is how I'm handling heterogeneity in DiscFac and CRRA. Assuming DiscFac count is 3 and CRRA count is 4, this is 12 distinct agent groupings. The next question is, how do we allow a parameter to vary over a CRRA subgroup but not a DiscFac subgroup. For example, let's assume each CRRA subgroup has different income process, so there are 4 different TranShkStd's, but there are 12 distinct agent subgroups. Do we repeat each TranShkStd 3 times in a DataArray?